### ISSUE TITLE
Compiler allows use of unassigned variable

### ISSUE THREAD

**Issue by dan655t** [dotnet/roslyn#51539]
**Version Used**: 
```
dotnet --version
5.0.100
```

**Steps to Reproduce**:

In the example below, the variable `myClass` can never be set.

```csharp
using System;

namespace console_app_test
{
    class Program
    {
        static void Main()
        {
            var val = new MyClass { Name = "Danny" };
            if (val is not MyClass myClass)
            {
                return;
            }
            Console.WriteLine(myClass.Name);
        }

        class MyClass
        {
            public string Name { get; set; }
        }
    }
}

```

Note, the compiler does catch this when the `return` is removed from the if block.

```csharp
using System;

namespace console_app_test
{
    class Program
    {
        static void Main()
        {
            var val = new MyClass { Name = "Danny" };
            if (val is not MyClass myClass)
            {
                // return;
            }
            Console.WriteLine(myClass.Name);
        }

        class MyClass
        {
            public string Name { get; set; }
        }
    }
}

```
`error CS0165: Use of unassigned local variable 'myClass'`

**Expected Behavior**:

A compiler error.

**Actual Behavior**:

Use of an unassigned variable compiles.

**Comment by RichardD012** [comment]
For what it's worth, I've reproduced this in the latest `5.0.103` version of the SDK as well.

In the above example, the `myClass` variable from the pattern matching `is not` is never set, even when the `val` in the expression is of type `MyClass`.

**Comment by RikkiGibson** [comment]
It's expected that for an expression like `val is not MyClass myClass`, the variable `myClass` is "definitely assigned when false". In other words, the behavior you're seeing appears to me to be by-design. IIRC @CyrusNajmabadi knows the motivation/context for this.

**Comment by RichardD012** [comment]
So based on this, the assignment to `myClass` can never happen when using `is not`? Seems like a weird language feature to declare a variable that it would be cast-to.  Sorry, your wording of "definitely assigned when false" doesn't seem to be the case.

In the above code, if you execute it, you get a NullPointerException on the `Console.WriteLine(myClass.Name);` line if the return is included.

https://dotnetfiddle.net/ZBTm3Z

In addition to this, if you move the condition to the outside of the if block, you get the compiler warning.  The follow code block won't compile.

```
	public static void Main()
	{
		var val = new MyClass { Name = "Danny" };
		var result = val is not MyClass myClass;
		if (result)
		{
			return;
		}
		Console.WriteLine(myClass.Name);
	}
```
The following compiles (and results in a Null ref).
```
	public static void Main()
	{
		var val = new MyClass { Name = "Danny" };
		if (val is not MyClass myClass)
		{
			return;
		}
		Console.WriteLine(myClass.Name);
	}
```

That doesn't seem by design.

**Comment by RikkiGibson** [comment]
I think there's a bug with this scenario in 16.8 that we fixed in 16.9. See #49369

It appears the scenario you've described has the expected behavior in SharpLab with the [master](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOyQExAagB8BiTdAG3ImHLgAI5NraBYAKAAEAmARnfYBuEBHSHk6AXjqY4AdzoBZAJ4BhKlCh0A3nQByEMPSkAiACIRMmJcboBfANzsAlgDM6ACjF0nmzAHsYRVV1TTBg6CgASnYtdjp4ug4Adkc2W3YOHgBOdzC1CIA6fUNI1IyAZkSuIPyNGLiEjkrMgAY9A3odAHM4GHs6KF7+9LT2IA===) branch but not with the [Default](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwMRQgG0wQwEaYCmABIVPkQLABQAAgEwCMNNAbjmMe5sQLzFRCAd2IBZAJ4BhXAGcZxAN7EAcjgC2JfgCIAIjihRxW4gF8A3DQCWAM2IAKbsUvyoAe2Bips+Wq845AJQ0CjTEYcS0AOwW1CY0tIwAnHa+0v4yAHSqGgEx8QDMEfSeaXLBoeG0hQkADCrqJEoA5oTAZsQyre1xsTRAA=).

**Comment by jaredpar** [comment]
The issue that was fixed for 16.9 is #49369

The behavior being described here though, ability to access the `myClass` in the `false` branch, is by design.
